File: cernael-python/4b.py
import re
def solve(lines):
    passports = []
    passport = {}
    for i in range(len(lines)):
        if ord(lines[i][0]) == 10:
            passports.append(passport)
            passport = {}
        else:
            fields = lines[i].split()
            for f in range(len(fields)):
                (k,v) = fields[f].split(':')
                passport[k] = v
    passports.append(passport)

    valid = 0
    for i in range(len(passports)):
        p = passports[i]
        if ('byr' in p and
            'iyr' in p and
            'eyr' in p and
            'hgt' in p and
            'hcl' in p and
            'ecl' in p and
            'pid' in p
        ):
            hcl_re = re.compile('^#[a-f0-9]{6}$')
            pid_re = re.compile('^[0-9]{9}$')
            if (
                1920 <= int(p['byr']) <= 2002 and 
                2010 <= int(p['iyr']) <= 2020 and 
                2020 <= int(p['eyr']) <= 2030 and 
                (
                 (p['hgt'][-2:] == 'in' and
                  59 <= int(p['hgt'][0:-2]) <= 76
                 ) or
                 (p['hgt'][-2:] == 'cm' and
                  150 <= int(p['hgt'][0:-2]) <= 193
                 )
                ) and
                hcl_re.match(p['hcl']) and
                p['ecl'] in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'] and
                pid_re.match(p['pid'])                
            ):
                valid += 1
    return valid
if __name__ == '__main__':
    lines = []
    with open('4.txt') as f:
        for line in f.readlines():
            lines.append(line)
    print(solve(lines))

File: nerarith-haskell/day4/4b.hs
import Data.List.Split (splitOn)
import qualified Data.Map as Map
import Data.Char (isDigit, isHexDigit)

main = interact (output . map (isValid . readInput) . splitOn [""] . lines)

readInput :: [String] -> Map.Map String String
readInput = Map.fromList . map splitPair . concatMap words

splitPair :: String -> (String, String)
splitPair s = (key, value)
    where (key:value:_) = splitOn ":" s

isValid :: Map.Map String String -> Bool
isValid d = and $ map check [(byr, "byr"), (iyr, "iyr"), (eyr, "eyr"), (hgt, "hgt"), (hcl, "hcl"), (ecl, "ecl"), (pid, "pid")]
    where check (f, key) = (Just True ==) . fmap f $ Map.lookup key d

byr = and . sequenceA [(>=1920), (<=2002)] . read
iyr = and . sequenceA [(>=2010), (<=2020)] . read
eyr = and . sequenceA [(>=2020), (<=2030)] . read
hgt = isValidLength . break (not . isDigit)
    where isValidLength (lenStr, unit)
            | null lenStr = False
            | unit == "cm" && 150 <= len && len <= 193 = True
            | unit == "in" && 59 <= len && len <= 76 = True
            | otherwise = False
            where len = read lenStr
hcl (h:number) = (h == '#') && (and $ map isHexDigit number)
ecl colour = elem colour ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
pid = and . sequenceA [((9 ==) . length), (and . map isDigit)]

output :: [Bool] -> String
output = (++"\n") . show . length . filter (True==)
